### PR TITLE
Tips and tricks

### PR DESCRIPTION
**Pull request by ITCups**
Subsystem for displaying tips and tricks
Added 2 verbs to display tips (first prints random tip, second print tip somehow concern character)
Tips automatically displayed 10 second after roundstart vote end, on antagonist role assigment (role specific tip) and on login on some mobs
default

### open poll for tips and tricks

### CODE REVIEW

**Comment by NanakoAC** [comment]
As an ironhammer operative, you have broad access to chase criminals. This does not mean you can take anything you have access to. Taking things from other departments is theft!

As a guild merchant, you should be buying up valueable things from scavengers and reselling them for a profit. Buy anything of value that's brought to you

If you're trying to take on a blob, a flamethrower is your best weapon.

The Fuhrer roach can call for an entire battalion of reinforcement roaches. If you hear it scream, RUN!

Spiders are a powerful foe that can poison or tranquilize their targets. Fighting them alone is a bad idea

Always carry a flashlight and some spare power cells. Darkness is dangerous

You can find or buy many modifications to improve your tools, and turn them into effective weapons

**Comment by Wesreidau** [comment]
As I started writing these I realized you could also make a number of them lore tidbits as well.

Antag

Roleplaying makes for loyal friends and respectful adversaries. Everyone loves a good storyteller who brings others into the plot.

Play to have fun and to bring others into the fun. If your round feels less like a kill compilation and more like a scifi thriller, you're doing it right.

Merely killing your enemies with a shot in the back is so very droll. You can do better. Make them die tired.

Being an antagonist is no excuse for not roleplaying.

Captain

A good leader's orders are always obeyed because a good leader never gives an order that won't be obeyed.

You didn't pick this role for the ID, did you?

This ship is the culmination of your life's work. Don't abandon it because you scratched the paint.

You have a vital out-of-character role; you are the quest-giver. If play lacks direction and the crew is getting restless, give them something productive to do.

The Captain is always right, even when he is wrong. Because they better hope you're right when you read that fuel gauge twelve light-days from port.

When do you break out the lash and airlock the ringleaders? Black. Bloody. Mutiny.

Technomancer / Technomancer Exult

In the ancient times, a group of engineers was called a tribe. That is why some technomancers call you Chief.

The Supreme Matter is a fickle mistress, and each Exultant has their own unique ritual. No ritual is strictly wrong; such a poor lover would be dead by now, surely...

Venting the Supreme Matter's overheated coolant may drop temperatures immediately, but without fresh coolant gas the Supreme Matter will almost surely delaminate.

Woe betide the Exultant who jettisons a Supreme Matter, one of the ancient relics of the grand, lost fleets; for their numbers dwindle, and the artifice of their creation is lost. It is better to hurl oneself after it than live so ashamed.

You are responsible for keeping these hundred souls alive inside an electrified air canister hurtling through an unholy blackness at speeds that make an Ironhammer bullet-pusher stare slack-jawed. You will fail. The question is how many bodies need fill the breach, smother the flames, and bind the wires.

A technomancer shall die that the ship shall live. It is the one rite all clans share.

The relationship between ship captains and technomancers is like that between tenants and landlords. One struts around between carpeting and ceiling like they own the place, while the other crawls in the foundation and attic actually giving a damn.

In case of fire, break glass.

No technomancer lives long without learning that firesuits are not space suits, and vice versa.

File: code/modules/tips_and_tricks/mobs.dm [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// This file contains all mobs tips and tricks and shown when you spawn or your mind is transfered
+/tipsAndTricks/mobs
+    var/list/mobs_list       //list of mobs to which tip can be shown
+    textColor = "#957820"
+
+/tipsAndTricks/mobs/breathing
+    mobs_list = list(/mob/living/carbon/human)
+    tipText = "As a human you need oxygen and a pressure to breath. Comfortable pressure for human is around 101psi. Some gasses like phoron are toxic to you."
```
Comment (Anonmare) [inline code/modules/tips_and_tricks/mobs.dm]: >Phoron
:angry:

Comment (comma) [reply code/modules/tips_and_tricks/mobs.dm]: Also kPa and psi mixup, 697 kPa isn't very comfortable.

**Comment by Anonmare** [comment]
Spiders and roaches can be stunned and slipped like humans, when facing off against either, try to even the fight by either flashing them with a flash or leading them to slip on some spilled water.
It's better to retreat from a fight and only suffer mild injuries than to win a fight and suffer a mortal wound, pick and choose your battles carefully.
Inaprovaline and dylovene can be mixed together in a container to produce tricordazine - a mild regenerative compound that can treat brute/burns/toxin damage and even suffocation, best of all, it has no overdose risk.
Merely applying gauze and ointment to your injuries won't fix the underlying damage or any compounding issues said injuries have caused. In addition, it is usually pointless to gauze a wound that isn't bleeding.
Pills containing pure carbon can be ingested to treat cases of poisoning and accidental overdoes, 1u of carbon will remove 1u of anything in the stomach. This will not, however, purge chemicals in the bloodstream, nor will injecting carbon have the same effect.
Inaprovaline and bicardine can be used to stem the effects of internal bleeding.

File: code/datums/outfits/jobs/security.dm [hunk 2]
```diff
@@ -60,10 +60,8 @@
 	uniform = /obj/item/clothing/under/rank/medspec
 	suit = /obj/item/clothing/suit/storage/toggle/labcoat/medspec
 	pda_type = /obj/item/modular_computer/pda/forensics
-	belt = /obj/item/weapon/gun/projectile/olivaw
 	l_hand = /obj/item/weapon/storage/briefcase/crimekit
-	backpack_contents = list(/obj/item/ammo_magazine/cl32/rubber = 2, /obj/item/weapon/storage/box/evidence = 1)
-
+	backpack_contents = list(/obj/item/ammo_magazine/cl32/rubber = 2, /obj/item/weapon/storage/box/evidence = 1, /obj/item/weapon/gun/projectile/olivaw = 1)
```
Comment (ACCount12) [inline code/datums/outfits/jobs/security.dm]: Why changes in this file?

File: code/controllers/subsystems/tips_manager.dm [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+/datum/controller/subsystem/tips/proc/getGameplayTip(var/startText)
+	if(GLOB.gameplayTips)
+		var/tipsAndTricks/T = pick(GLOB.gameplayTips)
+		return T
+
+/datum/controller/subsystem/tips/proc/getRoleTip(var/datum/antagonist/role)
+	if(!istype(role))
+		error("Not role type variable was passed to tips subsystem. No tips for you.")
+	var/list/tipsAndTricks/candidates = list()
+	log_world("giving tip for [role.id]")
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Debug message?

File: code/controllers/subsystems/tips_manager.dm [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	for(var/type in GLOB.rolesTips)
+		log_world("type [type]")
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: More debug messages.

File: code/controllers/subsystems/tips_manager.dm [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	wait = 60 MINUTES //Ticks once per 60 minute
+
+/client/verb/showRandomTip()
+	set name = "Show Random Tip"
+	set category = "OOC"
+	if(SSticker.current_state > GAME_STATE_STARTUP)
+		if(mob)
+			var/tipsAndTricks/T = SStips.getRandomTip()
+			if(T)
+				mob << SStips.formatTip(T, "Random Tip: ")
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `to_chat`.

File: code/controllers/subsystems/tips_manager.dm [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+			if(T)
+				mob << SStips.formatTip(T, "Tip for your character: ")
+
+/datum/controller/subsystem/tips/fire()
+	for(var/mob/living/L in SSmobs.mob_list)
+		if(L.client)
+			L.client.showSmartTip()
+
+/datum/controller/subsystem/tips/Initialize(start_timeofday)
+	for(var/path in typesof(/tipsAndTricks/mobs) - /tipsAndTricks/mobs)
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `subtypesof`.

File: code/controllers/subsystems/tips_manager.dm [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		var/tipsAndTricks/mobs/T = new path()
+		for(var/mob in T.mobs_list)
+			if(!GLOB.mobsTips[mob])
+				GLOB.mobsTips[mob] = list()
+			if(!GLOB.mobsTips[mob].Find(T))
+				GLOB.mobsTips[mob] += T
+	for(var/path in typesof(/tipsAndTricks/roles) - /tipsAndTricks/roles)
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `subtypesof`.

File: code/controllers/subsystems/tips_manager.dm [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+		var/tipsAndTricks/roles/T = new path()
+		for(var/role in T.roles_list)
+			if(!GLOB.rolesTips[role])
+				GLOB.rolesTips[role] = list()
+			if(!GLOB.rolesTips[role].Find(T))
+				GLOB.rolesTips[role] += T
+	for(var/path in typesof(/tipsAndTricks/jobs) - /tipsAndTricks/jobs)
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `subtypesof`.

File: code/controllers/subsystems/tips_manager.dm [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+		var/tipsAndTricks/jobs/T = new path()
+		for(var/job in T.jobs_list)
+			if(!GLOB.jobsTips[job])
+				GLOB.jobsTips[job] = list()
+			if(!GLOB.jobsTips[job].Find(T))
+				GLOB.jobsTips[job] += T
+	for(var/path in typesof(/tipsAndTricks/gameplay) - /tipsAndTricks/gameplay)
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `subtypesof`.

File: code/controllers/subsystems/vote.dm [hunk 10]
```diff
@@ -442,6 +442,10 @@ SUBSYSTEM_DEF(vote)
 		round_progressing = TRUE
 		world << "<b>The game will start in [SSticker.pregame_timeleft] seconds.</b>"
 	pregame = FALSE
+	spawn(10 SECONDS)
+		var/tipsAndTricks/T = SStips.getRandomTip()
+		if(T)
+			world << SStips.formatTip(T, "Random Tip: ")
```
Comment (ACCount12) [inline code/controllers/subsystems/vote.dm]: Use `to_chat`.

File: code/game/antagonist/antagonist_print.dm [hunk 11]
```diff
@@ -29,9 +29,15 @@
...
 	show_objectives()
-
+	printTip()
 	return TRUE
 
+/datum/antagonist/proc/printTip()
+	var/tipsAndTricks/T = SStips.getRoleTip(src)
+	if(T)
+		var/mob/player = owner.current
+		player << SStips.formatTip(T, "Tip for \a [src.id]: ")
```
Comment (ACCount12) [inline code/game/antagonist/antagonist_print.dm]: Use `to_chat`.

File: code/modules/tips_and_tricks/mobs.dm (hunk 1 above)
Comment (ACCount12) [reply code/modules/tips_and_tricks/mobs.dm]: Fix this.

File: code/controllers/subsystems/tips_manager.dm [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+				mob << SStips.formatTip(T, "Random Tip: ")
+
+/client/verb/showSmartTip()
+	set name = "Show Smart Tip"
+	set category = "OOC"
+	if(SSticker.current_state > GAME_STATE_STARTUP)
+		if(mob)
+			var/tipsAndTricks/T = SStips.getSmartTip(mob)
+			if(T)
+				mob << SStips.formatTip(T, "Tip for your character: ")
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: Use `to_chat`.

File: code/controllers/subsystems/tips_manager.dm [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+/datum/controller/subsystem/tips/proc/getSmartTip(var/mob/target)
+	if(!target)
+		return
+	// We need types
+	var/datum/antagonist/roleType
+	var/datum/job/jobType
+	if(target.mind)
+		roleType = target.mind.antagonist.len ? pick(target.mind.antagonist) : null	//pick random role cuz its a list
+		jobType = target.mind.assigned_job ? target.mind.assigned_job : null
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: What is the meaning of this statement?

File: code/controllers/subsystems/tips_manager.dm [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	var/mob/mobType = target ? target : null
```
Comment (ACCount12) [inline code/controllers/subsystems/tips_manager.dm]: What is the meaning of this statement?

**Pull request merged**